catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserver.java

```diff
@@ -0,0 +1,305 @@
+   * putting themselves in a state where they have never been committed.
+   *
+   * @param entry The AsyncFileEntry wrapping the file being listened to.
+   * @param success Boolean that shows if the task failed or completed successfully
+   */
+  private void commitDelete(AsyncFileEntry entry, boolean success) {
+    LOGGER.debug("commitDelete({},{}): Starting...", entry.getName(), success);
+    if (success) {
+      entry.getParent().ifPresent(e -> e.removeChild(entry));
+      entry.destroy();
+    }
+    removeFromProcessors(entry);
+  }
+
+  /**
+   * Steps file by file comparing the snapshot state to the current state of the directory being
+   * monitored.
+   *
+   * @param parent The parent directory (Wrapped in a AsyncFileEntry)
+   * @param previous The list of all children of the parent directory (In sorted order)
+   * @param files The list of current files (in sorted order)
+   */
+  private void checkAndNotify(
+      final AsyncFileEntry parent, final List<AsyncFileEntry> previous, final File[] files) {
+    int c = 0;
+    for (final AsyncFileEntry entry : previous) {
+      while (c < files.length && entry.compareToFile(files[c]) > 0) {
+        doCreate(new AsyncFileEntry(parent, files[c]));
+        c++;
+      }
+      if (c < files.length && entry.compareToFile(files[c]) == 0) {
+        doMatch(entry);
+        checkAndNotify(entry, entry.getChildren(), listFiles(files[c]));
+        c++;
+      } else {
+        //  Do Delete
+        if (!entry.checkNetwork()) {
+          //  The file MAY still exist but it's the network that's down.
+          return;
+        }
+        checkAndNotify(entry, entry.getChildren(), FileUtils.EMPTY_FILE_ARRAY);
+        doDelete(entry);
+      }
+    }
+    for (; c < files.length; c++) {
+      doCreate(new AsyncFileEntry(parent, files[c]));
+    }
+  }
+
+  /**
+   * Called when the observer should compare the snapshot state to the actual state of the directory
+   * being monitored.
+   */
+  public void checkAndNotify() {
+
+    if (listener == null) {
+      return;
+    }
+
+    /* fire directory/file events */
+    final File root = rootFile.getFile();
+    if (root.exists()) {
+      checkAndNotify(rootFile, rootFile.getChildren(), listFiles(rootFile.getFile()));
+    } else {
+      //  The file doesn't exist.
+      //  We assume this can't happen so this must be a network disconnect.
+      LOGGER.debug(
+          "The monitored file [{}] does not exist. No file processing will be done through the CDM",
+          rootFile.getName());
+    }
+  }
+
+  /**
+   * Note: returns a new Array to avoid sync access exceptions
+   *
+   * @return A new sorted File Array
+   */
+  private File[] listFiles(File file) {
+    if (file.isDirectory()) {
+      File[] temp = file.listFiles();
+      if (temp != null) {
+        Arrays.sort(temp);
+        return temp;
+      }
+    }
+    return FileUtils.EMPTY_FILE_ARRAY;
```

Comment:
❓ If there is an IO Error here and empty file array is returned, could we potentially end up deleting entries from the catalog if we still have connection to the NFS?

Reply:
Good catch, If it encounters an IO error should it just stop processing entirely until the next poll or just skip that entry and try to move to the next? I'm leaning to the ladder.